src/main/proto/spine/text/text.proto

```diff
@@ -43,15 +43,39 @@ message Text {
 // A position in the text.
 message Position {
 
-    // A zero-based index of the line of text.
-    //
-    // A value of -1 means "Not found".
-    //
-    int32 line = 1 [(min).value = "-1"];
+    oneof marker {
+
+        // A zero-based index of the line of text.
+        //
+        // A value of -1 means "Not found".
+        //
+        int32 line = 1 [(min).value = "-1"];
```

Comment:
So, `0`-s of both line and column would mean the same as one of the options in `beyond_text`? If so, I would simplify the latter by removing the "before all" option from it.

Basically, it would go like this:

- either both line and column are set to some non-negative values (with zeroes meaning "before all"),
- or the position is specified to be "after all".

Reply:
Then we probably also need to change validation restriction to allow zero as the minimum, and `not_in_text` being `boolean` under `oneof`.

So there are four cases for this type:
  * We look for existing text. It's either 1) there or 2) not.
  * We look for the place to add some text. It's either in 3) the text or 4) beyond it.

Reply:
Maybe we need to have a separate type for it, like `InsertionPoint`? 

It's either in the text, then it comes with the 1) `position`, or 2) beyond the text. 

Then it comes with a separate type with negative values for lines, but not columns (because we don't want to shift all the text because of the insertion).

Reply:
Or, maybe we need to have a type, like `Offset`, which allows both positive and negative values, so that it's easier to create relative points, instead of calculating a positive position from the very beginning of a line.

For example, I found a fully qualified type name, then I want to shift to the left using short name. I don't want to mess with counting from zero. I want to move the caret to the left. Like a typewriter or a Turing thing, you know?

Reply:
In any case, I suggest to introduce a performance test which would process about 200 Proto files with several
messages in each (and Java code produced from those). We will be able to understand where we stand in terms of processing/parsing speed before and after each change we make.

Reply:
200 is an approximate number of Proto files we have in `core-java/server/.../test` sources. Alternatively, we could count the number of messages declared there, and recreate this amount in scope of our testing.